File: apps/system/api/domain/me/received_invitation.go
package me

import (
	"github.com/ryo034/react-go-template/apps/system/api/domain/workspace"
	"github.com/ryo034/react-go-template/apps/system/api/domain/workspace/invitation"
)

// ReceivedInvitation is only active verified invitation
type ReceivedInvitation struct {
	invitation *invitation.Invitation
	inviter    workspace.Inviter
}

func NewReceivedInvitation(inv *invitation.Invitation, inviter workspace.Inviter) (ReceivedInvitation, error) {
	if inv.IsRevoked() {
		return ReceivedInvitation{}, invitation.NewAlreadyRevokedInvitation(inv.ID(), inv.Token().Value())
	}
	if inv.IsExpired() {
		return ReceivedInvitation{}, invitation.NewAlreadyExpiredInvitation(inv.ID(), inv.Token().Value())
	}
	if inv.IsAccepted() {
		return ReceivedInvitation{}, invitation.NewAlreadyAcceptedInvitation(inv.ID(), inv.Token().Value())
	}
	return ReceivedInvitation{inv, inviter}, nil
}

func (r *ReceivedInvitation) Invitation() *invitation.Invitation {
	return r.invitation
}

func (r *ReceivedInvitation) Inviter() workspace.Inviter {
	return r.inviter
}

File: apps/system/client/src/domain/me/receivedInvitation.ts
import { Invitation, Inviter } from "~/domain"
import { Entity } from "~/domain/shared"

interface Props {
  invitation: Invitation
  inviter: Inviter
}

export class ReceivedInvitation extends Entity<Props> {
  static create(v: Props): ReceivedInvitation {
    return new ReceivedInvitation(v)
  }

  get invitation(): Invitation {
    return this.value.invitation
  }

  get inviter(): Inviter {
    return this.value.inviter
  }
}
